ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,284 @@
+    pub async fn process_next_batch(
+        &mut self,
+        db: DbReaderWriter,
+    ) -> Vec<Result<EndVersion, Status>> {
+        let mut tasks = vec![];
+        let batches = self.get_batches().await;
+        let db_writer = db.writer.clone();
+
+        for batch in batches {
+            let start_time = std::time::Instant::now();
+            let db_writer = db_writer.clone();
+            let context = self.context.clone();
+            let ledger_version = self.highest_known_version;
+            let task = tokio::spawn(async move {
+                let raw_txns =
+                    Self::fetch_raw_txns_with_retries(context.clone(), ledger_version, batch).await;
+                Self::parse_table_info(context.clone(), raw_txns.clone(), db_writer.clone())
+                    .expect("Failed to parse table info");
+
+                DURATION_IN_SECS
+                    .with_label_values(&[
+                        SERVICE_TYPE,
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_step(),
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_label(),
+                    ])
+                    .set(start_time.elapsed().as_secs_f64());
+                TOTAL_SIZE_IN_BYTES
+                    .with_label_values(&[
+                        SERVICE_TYPE,
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_step(),
+                        IndexerTableInfoStep::TableInfoProcessedBatch.get_label(),
+                    ])
+                    .set(bcs::serialized_size(&raw_txns).unwrap() as i64);
+
+                Ok(raw_txns.last().unwrap().version)
+            });
+            tasks.push(task);
+        }
+        match futures::future::try_join_all(tasks).await {
+            Ok(res) => {
+                let db_reader = db.reader.clone();
+                let mut retried: u64 = 0;
+                while !db_reader
+                    .clone()
+                    .is_indexer_async_v2_pending_on_empty()
+                    .unwrap()
+                {
+                    retried += 1;
+                    let retry_batch = TransactionBatchInfo {
+                        start_version: self.current_version,
+                        num_transactions_to_fetch: self.parser_batch_size * self.parser_task_count,
+                    };
+                    let context = self.context.clone();
+                    let ledger_version = self.highest_known_version;
+                    let raw_txns = Self::fetch_raw_txns_with_retries(
+                        context.clone(),
+                        ledger_version,
+                        retry_batch,
+                    )
+                    .await;
+                    Self::parse_table_info(context.clone(), raw_txns.clone(), db_writer.clone())
+                        .expect("Failed to parse table info");
```

Comment:
End early if pending_on is empty.

Reply:
after putting more thoughts into it, i don't think there's way to know `pending_on` is truly empty unless we reach the end. in the mid of the processing, its always possible that `pending_on` gets to be empty when handle is being parsed, but might get added back if it fails to process.